File: bench/blur/blur.go
package main

import (
	"fmt"
	"time"
)

const (
	w       = 512
	h       = 512
	r       = 4
	win     = 2*r + 1
	n       = w * h * 4
	nRuns   = 21
	nWarmup = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func checksumU8(out []uint8) uint32 {
	hh := uint32(0x811c9dc5)
	for _, b := range out {
		hh = mix(hh, uint32(b))
	}
	return hh
}

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

func mkImage(out []uint8) {
	s := uint32(0x1234abcd)
	for i := range out {
		s ^= s << 13
		s ^= s >> 17
		s ^= s << 5
		out[i] = uint8(s & 255)
	}
}

func hblur(src, dst []uint8, w, h, r int) {
	win := 2*r + 1
	for y := 0; y < h; y++ {
		row := y * w
		for x := 0; x < w; x++ {
			sr, sg, sb, sa := 0, 0, 0, 0
			for k := -r; k <= r; k++ {
				xi := x + k
				if xi < 0 {
					xi = 0
				} else if xi >= w {
					xi = w - 1
				}
				p := (row + xi) << 2
				sr += int(src[p])
				sg += int(src[p+1])
				sb += int(src[p+2])
				sa += int(src[p+3])
			}
			o := (row + x) << 2
			dst[o] = uint8(sr / win)
			dst[o+1] = uint8(sg / win)
			dst[o+2] = uint8(sb / win)
			dst[o+3] = uint8(sa / win)
		}
	}
}

func vblur(src, dst []uint8, w, h, r int) {
	win := 2*r + 1
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			sr, sg, sb, sa := 0, 0, 0, 0
			for k := -r; k <= r; k++ {
				yi := y + k
				if yi < 0 {
					yi = 0
				} else if yi >= h {
					yi = h - 1
				}
				p := (yi*w + x) << 2
				sr += int(src[p])
				sg += int(src[p+1])
				sb += int(src[p+2])
				sa += int(src[p+3])
			}
			o := (y*w + x) << 2
			dst[o] = uint8(sr / win)
			dst[o+1] = uint8(sg / win)
			dst[o+2] = uint8(sb / win)
			dst[o+3] = uint8(sa / win)
		}
	}
}

func main() {
	img := make([]uint8, n)
	tmp := make([]uint8, n)
	out := make([]uint8, n)
	mkImage(img)
	for i := 0; i < nWarmup; i++ {
		hblur(img, tmp, w, h, r)
		vblur(tmp, out, w, h, r)
	}
	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		hblur(img, tmp, w, h, r)
		vblur(tmp, out, w, h, r)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), checksumU8(out), w*h, win, nRuns)
}

File: bench/blur/blur.rs
use std::time::Instant;

const W: i32 = 512;
const H: i32 = 512;
const R: i32 = 4;
const WIN: i32 = 2 * R + 1;
const N: usize = (W * H * 4) as usize;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn checksum_u8(out: &[u8]) -> u32 {
    let mut h = 0x811c_9dc5u32;
    for &b in out {
        h = mix(h, b as u32);
    }
    h
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn mk_image(out: &mut [u8]) {
    let mut s = 0x1234abcdu32;
    for x in out.iter_mut() {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        *x = (s & 255) as u8;
    }
}

fn hblur(src: &[u8], dst: &mut [u8], w: i32, h: i32, r: i32) {
    let win = 2 * r + 1;
    for y in 0..h {
        let row = y * w;
        for x in 0..w {
            let (mut sr, mut sg, mut sb, mut sa) = (0i32, 0i32, 0i32, 0i32);
            for k in -r..=r {
                let mut xi = x + k;
                if xi < 0 {
                    xi = 0;
                } else if xi >= w {
                    xi = w - 1;
                }
                let p = ((row + xi) << 2) as usize;
                sr += src[p] as i32;
                sg += src[p + 1] as i32;
                sb += src[p + 2] as i32;
                sa += src[p + 3] as i32;
            }
            let o = ((row + x) << 2) as usize;
            dst[o] = (sr / win) as u8;
            dst[o + 1] = (sg / win) as u8;
            dst[o + 2] = (sb / win) as u8;
            dst[o + 3] = (sa / win) as u8;
        }
    }
}

fn vblur(src: &[u8], dst: &mut [u8], w: i32, h: i32, r: i32) {
    let win = 2 * r + 1;
    for y in 0..h {
        for x in 0..w {
            let (mut sr, mut sg, mut sb, mut sa) = (0i32, 0i32, 0i32, 0i32);
            for k in -r..=r {
                let mut yi = y + k;
                if yi < 0 {
                    yi = 0;
                } else if yi >= h {
                    yi = h - 1;
                }
                let p = ((yi * w + x) << 2) as usize;
                sr += src[p] as i32;
                sg += src[p + 1] as i32;
                sb += src[p + 2] as i32;
                sa += src[p + 3] as i32;
            }
            let o = ((y * w + x) << 2) as usize;
            dst[o] = (sr / win) as u8;
            dst[o + 1] = (sg / win) as u8;
            dst[o + 2] = (sb / win) as u8;
            dst[o + 3] = (sa / win) as u8;
        }
    }
}

fn main() {
    let mut img = vec![0u8; N];
    let mut tmp = vec![0u8; N];
    let mut out = vec![0u8; N];
    mk_image(&mut img);
    for _ in 0..N_WARMUP {
        hblur(&img, &mut tmp, W, H, R);
        vblur(&tmp, &mut out, W, H, R);
    }
    let mut samples = [0.0; N_RUNS];
    for sample in &mut samples {
        let t0 = Instant::now();
        hblur(&img, &mut tmp, W, H, R);
        vblur(&tmp, &mut out, W, H, R);
        *sample = t0.elapsed().as_secs_f64() * 1000.0;
    }
    let us = median_us(&mut samples);
    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        us,
        checksum_u8(&out),
        W * H,
        WIN,
        N_RUNS
    );
}
